samcli/lib/sync/flows/generic_api_sync_flow.py

```diff
@@ -65,19 +65,24 @@ def gather_resources(self) -> None:
     def _process_definition_file(self) -> Optional[bytes]:
         if self._definition_uri is None:
             return None
-        with open(self._definition_uri, "rb") as swagger_file:
+        with open(str(self._definition_uri), "rb") as swagger_file:
             swagger_body = swagger_file.read()
             return swagger_body
 
-    def _get_definition_file(self, api_identifier: str) -> Optional[str]:
+    def _get_definition_file(self, api_identifier: str) -> Optional[Path]:
         api_resource = get_resource_by_id(self._stacks, ResourceIdentifier(api_identifier))
         if api_resource is None:
             return None
         properties = api_resource.get("Properties", {})
         definition_file = properties.get("DefinitionUri")
-        if self._build_context.base_dir and definition_file:
-            definition_file = str(Path(self._build_context.base_dir).joinpath(definition_file))
-        return cast(Optional[str], definition_file)
+        definition_path = None
+        if definition_file:
+            definition_path = Path(self._build_context.base_dir).joinpath(definition_file)
+            if not self._build_context.use_base_dir:
+                child_stack = Stack.get_stack_by_full_path(ResourceIdentifier(api_identifier).stack_path, self._stacks)
+                if child_stack:
+                    definition_path = Path(child_stack.location).parent.joinpath(definition_file)
```

Comment:
This seems duplicated lines with SFN flow. Can we move this one into a common function so that we can re-use it in sfn sync flow as well?
Can we also add some debug logs if we need to investigate an issue in the future?